crates/rpc/tests/rpc.rs

```diff
@@ -0,0 +1,488 @@
+    let client = JsonRpcClient::new(HttpTransport::new(rpc_url));
+
+    let block = match client
+        .get_block_with_txs(BlockId::Tag(BlockTag::Latest))
+        .await
+        .expect("Failed to retrieve the latest block")
+    {
+        MaybePendingBlockWithTxs::Block(block) => block,
+        MaybePendingBlockWithTxs::PendingBlock(_) => {
+            panic!("The latest block shouldn't be a pending block")
+        }
+    };
+
+    if let Some(extracted_value) = extractor(&block) {
+        return TestContext {
+            client,
+            block_id: BlockId::Number(block.block_number),
+            block,
+            extracted_value,
+        };
+    }
+
+    // The latest block doesn't match the criterion.
+    // Go to the lower blocks one by one until a match is found.
+
+    let mut limit = 1_000;
+    let mut block_number = block.block_number;
+
+    while limit > 0 {
+        limit -= 1;
+        block_number -= 1;
+
+        if block_number == 0 {
+            panic!("Reached the genesis block, still no suitable block found");
+        }
+
+        let block = match client
+            .get_block_with_txs(BlockId::Number(block_number))
+            .await
+            .expect("Block retrieval failed")
+        {
+            MaybePendingBlockWithTxs::Block(block) => block,
+            MaybePendingBlockWithTxs::PendingBlock(_) => {
+                continue;
+            }
+        };
+
+        if let Some(extracted_value) = extractor(&block) {
+            return TestContext {
+                client,
+                block_id: BlockId::Number(block.block_number),
+                block,
+                extracted_value,
+            };
+        }
+    }
+
+    panic!("No suitable block found")
+}
+
+/// An extension trait providing extra methods to the Starknet client.
+pub trait L2ClientExt {
+    /// Retrieve the list of all the events of a single block.
+    fn get_block_events(
+        &self,
+        block_id: BlockId,
+    ) -> impl std::future::Future<Output = Result<Vec<Event>, ProviderError>> + Send;
+}
+
+impl L2ClientExt for JsonRpcClient<HttpTransport> {
+    async fn get_block_events(
+        &self,
+        block_id: BlockId,
+    ) -> Result<Vec<Event>, ProviderError> {
+        let page_size = 1024;
+        let mut events = vec![];
+        let mut last_token = None;
+        let mut more_to_fetch = true;
+
+        while more_to_fetch {
+            let events_page = self
+                .get_events(
+                    EventFilter {
+                        from_block: Some(block_id),
+                        to_block: Some(block_id),
+                        address: None,
+                        keys: None,
+                    },
+                    last_token,
+                    page_size,
+                )
+                .await?;
+            last_token = events_page.continuation_token;
+
+            if last_token.is_none() {
```

Comment:
`more_to_fetch` is always equal to `last_token.is_some()`, it can be substituted then such check is not necessary